### PR TITLE
Loading Notification

### PR DESCRIPTION
**Pull request by HunterBarclay**
### Description
Adds a loading notification to the bottom of the screen. This can be used via the `ProgressHandle` class which communicates to the UI via events. Progress handles can then be used to update status of tasks, and show completion or error when necessary.

### Tasks
- [x] Generalized `ProgressHandle` for UI interfacing outside of a react node.
- [x] Animate cards and progress bars.
- [x] Integrate with Mirabuf assembly spawning.

### Async Loading
So I've looked into how to get the loading to stop holding the main thread, and I haven't found a good option. Web workers have horrid restrictions on shared data and I haven't been able to find anything on deprioritizing the promise that loads the mirabuf assembly.

### Style
image
image

### CODE REVIEW

File: fission/src/mirabuf/MirabufParser.ts [hunk 1]
```diff
@@ -72,11 +73,13 @@ class MirabufParser {
...
     }
 
-    public constructor(assembly: mirabuf.Assembly) {
+    public constructor(assembly: mirabuf.Assembly, progressHandle?: ProgressHandle) {
         this._assembly = assembly
         this._errors = new Array<ParseError>()
         this._globalTransforms = new Map()
 
+        progressHandle?.Update("Parsing assembly...", 0.3)
+
```
Comment (Dhruv-0-Arora) [inline fission/src/mirabuf/MirabufParser.ts:82]: Would it be better to move this to MirabufSceneObject instead?

File: fission/src/ui/components/ProgressNotification.tsx [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+function Interp(elapse: number, progressData: ProgressData) {
+    const [value, setValue] = useState<number>(0)
+
+    useEffect(() => {
+        const update = () => {
+            const n = Math.min(1.0, Math.max(0.0, (Date.now() - progressData.lastUpdate) / elapse))
+            const v = progressData.lastValue + (progressData.currentValue - progressData.lastValue) * easeOutQuad(n)
+
+            setValue(v)
+        }
```
Comment (Dhruv-0-Arora) [inline fission/src/ui/components/ProgressNotification.tsx]: Could you add documentation as to what is happening here

File: fission/src/util/EasingFunctions.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+/**
+ * Source: https://easings.net/#easeOutQuad
+ *
+ * @param x
+ * @returns
```
Comment (Dhruv-0-Arora) [inline fission/src/util/EasingFunctions.ts]: Could you update this documentation as well

**Review by Dhruv-0-Arora — APPROVED** [review]
Overall looks good; it would be helpful to add documentation

File: fission/src/mirabuf/MirabufParser.ts (hunk 1 above)
Comment (HunterBarclay) [reply fission/src/mirabuf/MirabufParser.ts:82]: The point of that update is to indicate that the assembly is being parsed, so no.

Comment (HunterBarclay) [reply fission/src/mirabuf/MirabufParser.ts:82]: Also, the MirabufSceneObject doesn't exist at that point

**Review by LucaHaverty — APPROVED** [review]
Great work! Gives people something good to look at while a robot is loading instead of thinking Synthesis is frozen 👍👍👍

**Pull request merged**